From: Antony Dovgal <tony@daylessday.org>
Date: Fri, 11 Jul 2008 18:29:40 +0400
Subject: [PATCH] fix bug #0000002 (Ruby binding: hpdf_insert_page has stray printf)

---
 if/ruby/hpdf.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/if/ruby/hpdf.c b/if/ruby/hpdf.c
index 9191dc2..4617371 100644
--- a/if/ruby/hpdf.c
+++ b/if/ruby/hpdf.c
@@ -371,7 +371,6 @@ hpdf_insert_page (VALUE obj, VALUE page)
     Data_Get_Struct(obj, HPDF_Doc_Rec, pdf);
     Data_Get_Struct(page, HPDF_Dict_Rec, target);
 
-    printf("@page=%p\n", target);
     HPDF_PTRACE(("hpdf_insert_page pdf=%p\n", pdf));
 
     new_page = HPDF_InsertPage(pdf, target);
-- 
