From: Antony Dovgal <tony@daylessday.org>
Date: Sat, 11 Oct 2008 23:50:05 +0400
Subject: [PATCH] fix bug #0000005 (possible NULL derefernce in HPDF_LoadRawImageFromFile())

---
 src/hpdf_doc.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/hpdf_doc.c b/src/hpdf_doc.c
index 33d5a5c..64c8931 100644
--- a/src/hpdf_doc.c
+++ b/src/hpdf_doc.c
@@ -1628,7 +1628,7 @@ HPDF_LoadRawImageFromFile  (HPDF_Doc          pdf,
     if (!image)
         HPDF_CheckError (&pdf->error);
 
-    if (pdf->compression_mode & HPDF_COMP_IMAGE)
+    if (image && pdf->compression_mode & HPDF_COMP_IMAGE)
         image->filter = HPDF_STREAM_FILTER_FLATE_DECODE;
 
     return image;
@@ -1656,7 +1656,7 @@ HPDF_LoadRawImageFromMem  (HPDF_Doc           pdf,
     if (!image)
         HPDF_CheckError (&pdf->error);
 
-    if (pdf->compression_mode & HPDF_COMP_IMAGE)
+    if (image && pdf->compression_mode & HPDF_COMP_IMAGE)
         image->filter = HPDF_STREAM_FILTER_FLATE_DECODE;
 
     return image;
-- 
