From: Johan Van de Wauw <johan.vandewauw@gmail.com>
Date: Thu, 17 Jun 2010 09:53:31 +0200
Subject: [PATCH] Update example to work with font in debian

---
 demo/encoding_list.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/demo/encoding_list.c b/demo/encoding_list.c
index f93837c..faea48b 100644
--- a/demo/encoding_list.c
+++ b/demo/encoding_list.c
@@ -196,8 +196,8 @@ int main (int argc, char **argv)
     font_name = HPDF_LoadType1FontFromFile (pdf, "type1\\a010013l.afm",
             "type1\\a010013l.pfb");
     #else
-    font_name = HPDF_LoadType1FontFromFile (pdf, "type1/a010013l.afm",
-            "type1/a010013l.pfb");
+    font_name = HPDF_LoadType1FontFromFile (pdf, "/usr/share/fonts/type1/a010013l.afm",
+            "/usr/share/fonts/type1/a010013l.pfb");
     #endif
 
     /* create outline root. */
-- 
