From: Antony Dovgal <tony@daylessday.org>
Date: Mon, 30 Jun 2008 12:02:25 +0400
Subject: [PATCH] fix bug reported by chiisaitsu at gmail dot com

---
 if/ruby/hpdf.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/if/ruby/hpdf.c b/if/ruby/hpdf.c
index 3db5869..a021958 100644
--- a/if/ruby/hpdf.c
+++ b/if/ruby/hpdf.c
@@ -1388,7 +1388,7 @@ hpdf_set_info_date_attr (VALUE obj, VALUE type, VALUE year, VALUE month, VALUE d
 
     Data_Get_Struct(obj, HPDF_Doc_Rec, pdf);
     s = STR2CSTR(ind);
-    if (HPDF_StrLen(s, -1) != 1 || s[0] != '+' || s[0] != '-' || s[0] != 'Z')
+    if (HPDF_StrLen(s, -1) != 1 || (s[0] != '+' || s[0] != '-' || s[0] != 'Z'))
         rb_raise(rb_eHPDFError, "The eighth argument must be either of '+','-','Z',' '.");
 
     tp = NUM2INT(type);
-- 
