From: Antony Dovgal <tony@daylessday.org>
Date: Sat, 11 Oct 2008 23:28:42 +0400
Subject: [PATCH] fix bug #0000004 (possible NULL dereference in HPDF_AToI())

---
 src/hpdf_utils.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/src/hpdf_utils.c b/src/hpdf_utils.c
index 29114a3..1225caf 100644
--- a/src/hpdf_utils.c
+++ b/src/hpdf_utils.c
@@ -27,6 +27,10 @@ HPDF_AToI  (const char  *s)
     HPDF_BOOL flg = HPDF_FALSE;
     HPDF_INT  v = 0;
 
+    if (!s) {
+        return 0;
+    }
+
     /* increment pointer until the charactor of 's' is not
      * white-space-charactor.
      */
-- 
