From: Antony Dovgal <tony@daylessday.org>
Date: Tue, 2 Feb 2010 13:12:37 +0300
Subject: [PATCH] Fix missing parentheses from empty string object (Ralf Junker)

---
 src/hpdf_streams.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/src/hpdf_streams.c b/src/hpdf_streams.c
index aa0491d..f33e1e2 100644
--- a/src/hpdf_streams.c
+++ b/src/hpdf_streams.c
@@ -421,8 +421,14 @@ HPDF_Stream_WriteEscapeText2  (HPDF_Stream    stream,
 
     HPDF_PTRACE((" HPDF_Stream_WriteEscapeText2\n"));
 
+   /* The following block is commented out because it violates "PDF Spec 7.3.4.2 Literal Strings". 
+	* It states that the two matching parentheses must still be present to represent an empty 
+	* string of zero length. 
+	*/
+   /*
     if (!len)
         return HPDF_OK;
+   */
 
     buf[idx++] = '(';
 
-- 
